consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/pki/PkiQbftExtraDataCodec.java

```diff
@@ -20,71 +23,66 @@
 import org.hyperledger.besu.ethereum.rlp.BytesValueRLPOutput;
 import org.hyperledger.besu.ethereum.rlp.RLPInput;
 
-import java.util.Optional;
+import java.util.List;
 
 import org.apache.tuweni.bytes.Bytes;
 
 /*
- The PkiQbftExtraData encoding format is different from the "regular" QbftExtraData encoding.
- We have an "envelope" list, with two elements: the extra data and the cms message.
- The RLP encoding format is as follows: ["extra_data", ["cms"]]
+ The PkiQbftExtraData encoding format is different from the "regular" QbftExtraData encoding. We
+ have an extra bytes element in the end of the list.
 */
 public class PkiQbftExtraDataCodec extends QbftExtraDataCodec {
 
+  public static final int QBFT_EXTRA_DATA_LIST_SIZE = 5;
+
   @Override
   public BftExtraData decodeRaw(final Bytes input) {
     if (input.isEmpty()) {
       throw new IllegalArgumentException("Invalid Bytes supplied - Bft Extra Data required.");
     }
 
-    final RLPInput rlpInput = new BytesValueRLPInput(input, false);
-    rlpInput.enterList();
+    final BftExtraData bftExtraData = super.decodeRaw(input);
 
-    // Consume all the ExtraData input from the envelope list, and decode it using the QBFT decoder
-    final Bytes extraDataListAsBytes = rlpInput.currentListAsBytes();
-    final BftExtraData bftExtraData = super.decodeRaw(extraDataListAsBytes);
+    final RLPInput rlpInput = new BytesValueRLPInput(input, false);
 
-    final Optional<Bytes> cms;
-    if (rlpInput.nextIsList() && rlpInput.nextSize() == 0) {
-      cms = Optional.empty();
+    final Bytes cms;
+    final List<RLPInput> elements = rlpInput.readList(RLPInput::readAsRlp);
+    if (elements.size() > QBFT_EXTRA_DATA_LIST_SIZE) {
```

Comment:
is there a requirement or scenario to check the upper limit of elements? Is `5` size without cms?

Reply:
@jframe ^

Reply:
Unfortunately the upper limit comes from Qbft extra data codec. Don't see any real way around knowing what the expected number of items for qbft extra data is. Fortunately we have tests around this so if the qbft extra data rlp list size changes our pki qbft extra data tests will fail.